File: sdk/go/terminal.go
package agenttrace

import (
	"bytes"
	"context"
	"encoding/json"
	"os"
	"os/exec"
	"time"

	"github.com/google/uuid"
)

// TerminalCommandOptions holds options for tracking a terminal command.
type TerminalCommandOptions struct {
	Command          string
	Args             []string
	ObservationID    string
	WorkingDirectory string
	Shell            string
	EnvVars          map[string]string
	ExitCode         int
	Stdout           string
	Stderr           string
	StdoutTruncated  bool
	StderrTruncated  bool
	TimedOut         bool
	Killed           bool
	MaxMemoryBytes   int64
	CPUTimeMs        int64
	ToolName         string
	Reason           string
	StartedAt        *time.Time
	CompletedAt      *time.Time
	Success          *bool
}

// TerminalCommandInfo contains information about a tracked terminal command.
type TerminalCommandInfo struct {
	ID               string
	TraceID          string
	ObservationID    string
	Command          string
	Args             []string
	WorkingDirectory string
	ExitCode         int
	Stdout           string
	Stderr           string
	Success          bool
	DurationMs       int64
	StartedAt        time.Time
	CompletedAt      time.Time
}

// RunCommandOptions holds options for running and tracking a command.
type RunCommandOptions struct {
	Args             []string
	ObservationID    string
	WorkingDirectory string
	Env              map[string]string
	Timeout          time.Duration
	ToolName         string
	Reason           string
	MaxOutputBytes   int
}

// RunCommandResult contains the result of running a command.
type RunCommandResult struct {
	Info     *TerminalCommandInfo
	ExitCode int
	Stdout   string
	Stderr   string
}

// TerminalCmd tracks a terminal command for this trace.
func (t *Trace) TerminalCmd(opts TerminalCommandOptions) *TerminalCommandInfo {
	cmdID := uuid.New().String()
	now := time.Now().UTC()

	startedAt := now
	if opts.StartedAt != nil {
		startedAt = *opts.StartedAt
	}

	completedAt := now
	if opts.CompletedAt != nil {
		completedAt = *opts.CompletedAt
	}

	durationMs := completedAt.Sub(startedAt).Milliseconds()

	workingDirectory := opts.WorkingDirectory
	if workingDirectory == "" {
		workingDirectory, _ = os.Getwd()
	}

	args := opts.Args
	if args == nil {
		args = []string{}
	}

	success := opts.ExitCode == 0
	if opts.Success != nil {
		success = *opts.Success
	}

	// Convert env vars to JSON string
	var envVarsStr string
	if opts.EnvVars != nil {
		if data, err := json.Marshal(opts.EnvVars); err == nil {
			envVarsStr = string(data)
		}
	}

	// Send to API
	if t.client.Enabled() {
		t.client.addEvent(map[string]any{
			"type": "terminal-command-create",
			"body": map[string]any{
				"id":               cmdID,
				"traceId":          t.id,
				"observationId":    opts.ObservationID,
				"command":          opts.Command,
				"args":             args,
				"workingDirectory": workingDirectory,
				"shell":            opts.Shell,
				"envVars":          envVarsStr,
				"startedAt":        startedAt.Format(time.RFC3339Nano),
				"completedAt":      completedAt.Format(time.RFC3339Nano),
				"durationMs":       durationMs,
				"exitCode":         opts.ExitCode,
				"stdout":           opts.Stdout,
				"stderr":           opts.Stderr,
				"stdoutTruncated":  opts.StdoutTruncated,
				"stderrTruncated":  opts.StderrTruncated,
				"success":          success,
				"timedOut":         opts.TimedOut,
				"killed":           opts.Killed,
				"maxMemoryBytes":   opts.MaxMemoryBytes,
				"cpuTimeMs":        opts.CPUTimeMs,
				"toolName":         opts.ToolName,
				"reason":           opts.Reason,
			},
		})
	}

	return &TerminalCommandInfo{
		ID:               cmdID,
		TraceID:          t.id,
		ObservationID:    opts.ObservationID,
		Command:          opts.Command,
		Args:             args,
		WorkingDirectory: workingDirectory,
		ExitCode:         opts.ExitCode,
		Stdout:           opts.Stdout,
		Stderr:           opts.Stderr,
		Success:          success,
		DurationMs:       durationMs,
		StartedAt:        startedAt,
		CompletedAt:      completedAt,
	}
}

// RunCmd runs a command and tracks it.
func (t *Trace) RunCmd(ctx context.Context, command string, opts *RunCommandOptions) *RunCommandResult {
	if opts == nil {
		opts = &RunCommandOptions{}
	}

	startedAt := time.Now().UTC()

	workingDirectory := opts.WorkingDirectory
	if workingDirectory == "" {
		workingDirectory, _ = os.Getwd()
	}

	maxOutputBytes := opts.MaxOutputBytes
	if maxOutputBytes == 0 {
		maxOutputBytes = 100000
	}

	args := opts.Args
	if args == nil {
		args = []string{}
	}

	// Create command
	var cmd *exec.Cmd
	if opts.Timeout > 0 {
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(ctx, opts.Timeout)
		defer cancel()
	}

	cmd = exec.CommandContext(ctx, command, args...)
	cmd.Dir = workingDirectory

	// Set environment
	if opts.Env != nil {
		cmd.Env = os.Environ()
		for k, v := range opts.Env {
			cmd.Env = append(cmd.Env, k+"="+v)
		}
	}

	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()

	completedAt := time.Now().UTC()

	timedOut := ctx.Err() == context.DeadlineExceeded
	killed := cmd.ProcessState != nil && !cmd.ProcessState.Exited()

	exitCode := 0
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		} else {
			exitCode = -1
		}
	}

	stdoutStr := stdout.String()
	stderrStr := stderr.String()

	stdoutTruncated := len(stdoutStr) > maxOutputBytes
	stderrTruncated := len(stderrStr) > maxOutputBytes

	if stdoutTruncated {
		stdoutStr = stdoutStr[:maxOutputBytes]
	}
	if stderrTruncated {
		stderrStr = stderrStr[:maxOutputBytes]
	}

	info := t.TerminalCmd(TerminalCommandOptions{
		Command:          command,
		Args:             args,
		ObservationID:    opts.ObservationID,
		WorkingDirectory: workingDirectory,
		ExitCode:         exitCode,
		Stdout:           stdoutStr,
		Stderr:           stderrStr,
		StdoutTruncated:  stdoutTruncated,
		StderrTruncated:  stderrTruncated,
		TimedOut:         timedOut,
		Killed:           killed,
		ToolName:         opts.ToolName,
		Reason:           opts.Reason,
		StartedAt:        &startedAt,
		CompletedAt:      &completedAt,
	})

	return &RunCommandResult{
		Info:     info,
		ExitCode: exitCode,
		Stdout:   stdoutStr,
		Stderr:   stderrStr,
	}
}

File: sdk/typescript/src/terminal.ts
/**
 * Terminal command tracking for AgentTrace.
 *
 * Track terminal/shell commands executed during agent sessions for
 * full visibility and debugging capabilities.
 */

import { spawn, SpawnOptions } from "child_process";
import { cwd } from "process";
import type { AgentTrace } from "./client";

export interface TerminalCommandOptions {
  command: string;
  args?: string[];
  observationId?: string;
  workingDirectory?: string;
  shell?: string;
  envVars?: Record<string, string>;
  exitCode?: number;
  stdout?: string;
  stderr?: string;
  stdoutTruncated?: boolean;
  stderrTruncated?: boolean;
  timedOut?: boolean;
  killed?: boolean;
  maxMemoryBytes?: number;
  cpuTimeMs?: number;
  toolName?: string;
  reason?: string;
  startedAt?: Date;
  completedAt?: Date;
  success?: boolean;
}

export interface TerminalCommandInfo {
  id: string;
  traceId: string;
  observationId?: string;
  command: string;
  args: string[];
  workingDirectory: string;
  exitCode: number;
  stdout: string;
  stderr: string;
  success: boolean;
  durationMs: number;
  startedAt: Date;
  completedAt: Date;
}

export interface RunCommandOptions {
  args?: string[];
  observationId?: string;
  workingDirectory?: string;
  env?: Record<string, string>;
  timeout?: number;
  shell?: boolean | string;
  toolName?: string;
  reason?: string;
  maxOutputBytes?: number;
}

export interface RunCommandResult {
  info: TerminalCommandInfo;
  exitCode: number;
  stdout: string;
  stderr: string;
}

/**
 * Client for tracking terminal commands.
 *
 * @example
 * ```typescript
 * const client = new AgentTrace({ apiKey: "..." });
 * const trace = client.trace({ name: "my-agent" });
 *
 * // Track a command manually
 * trace.terminalCmd({
 *   command: "npm",
 *   args: ["test"],
 *   exitCode: 0,
 *   stdout: "All tests passed"
 * });
 *
 * // Or run and track automatically
 * const result = await trace.runCmd("npm", ["test"]);
 * ```
 */
export class TerminalClient {
  private _client: AgentTrace;

  constructor(client: AgentTrace) {
    this._client = client;
  }

  /**
   * Track a terminal command.
   */
  track(traceId: string, options: TerminalCommandOptions): TerminalCommandInfo {
    const cmdId = generateId();
    const now = new Date();

    const startedAt = options.startedAt || now;
    const completedAt = options.completedAt || now;
    const durationMs = completedAt.getTime() - startedAt.getTime();

    const workingDirectory = options.workingDirectory || cwd();
    const args = options.args || [];
    const stdout = options.stdout || "";
    const stderr = options.stderr || "";
    const exitCode = options.exitCode ?? 0;
    const success = options.success ?? (exitCode === 0);

    // Convert env vars to JSON string
    const envVarsStr = options.envVars ? JSON.stringify(options.envVars) : undefined;

    // Send to API
    if (this._client.enabled) {
      this._client._addEvent({
        type: "terminal-command-create",
        body: {
          id: cmdId,
          traceId,
          observationId: options.observationId,
          command: options.command,
          args,
          workingDirectory,
          shell: options.shell,
          envVars: envVarsStr,
          startedAt: startedAt.toISOString(),
          completedAt: completedAt.toISOString(),
          durationMs,
          exitCode,
          stdout,
          stderr,
          stdoutTruncated: options.stdoutTruncated || false,
          stderrTruncated: options.stderrTruncated || false,
          success,
          timedOut: options.timedOut || false,
          killed: options.killed || false,
          maxMemoryBytes: options.maxMemoryBytes,
          cpuTimeMs: options.cpuTimeMs,
          toolName: options.toolName,
          reason: options.reason,
        },
      });
    }

    return {
      id: cmdId,
      traceId,
      observationId: options.observationId,
      command: options.command,
      args,
      workingDirectory,
      exitCode,
      stdout,
      stderr,
      success,
      durationMs,
      startedAt,
      completedAt,
    };
  }

  /**
   * Run a command and track it.
   */
  async run(traceId: string, command: string, options: RunCommandOptions = {}): Promise<RunCommandResult> {
    const startedAt = new Date();
    const workingDirectory = options.workingDirectory || cwd();
    const maxOutputBytes = options.maxOutputBytes || 100000;

    return new Promise((resolve) => {
      const args = options.args || [];
      let stdout = "";
      let stderr = "";
      let timedOut = false;
      let killed = false;

      const spawnOptions: SpawnOptions = {
        cwd: workingDirectory,
        env: options.env ? { ...process.env, ...options.env } : process.env,
        shell: options.shell,
      };

      const proc = spawn(command, args, spawnOptions);

      let timeoutId: NodeJS.Timeout | undefined;
      if (options.timeout) {
        timeoutId = setTimeout(() => {
          timedOut = true;
          killed = true;
          proc.kill("SIGTERM");
        }, options.timeout);
      }

      proc.stdout?.on("data", (data: Buffer) => {
        if (stdout.length < maxOutputBytes) {
          stdout += data.toString().slice(0, maxOutputBytes - stdout.length);
        }
      });

      proc.stderr?.on("data", (data: Buffer) => {
        if (stderr.length < maxOutputBytes) {
          stderr += data.toString().slice(0, maxOutputBytes - stderr.length);
        }
      });

      proc.on("close", (code) => {
        if (timeoutId) clearTimeout(timeoutId);

        const completedAt = new Date();
        const exitCode = code ?? -1;

        const stdoutTruncated = stdout.length >= maxOutputBytes;
        const stderrTruncated = stderr.length >= maxOutputBytes;

        const info = this.track(traceId, {
          command,
          args,
          observationId: options.observationId,
          workingDirectory,
          exitCode,
          stdout,
          stderr,
          stdoutTruncated,
          stderrTruncated,
          timedOut,
          killed,
          toolName: options.toolName,
          reason: options.reason,
          startedAt,
          completedAt,
        });

        resolve({
          info,
          exitCode,
          stdout,
          stderr,
        });
      });

      proc.on("error", (error) => {
        if (timeoutId) clearTimeout(timeoutId);

        const completedAt = new Date();

        const info = this.track(traceId, {
          command,
          args,
          observationId: options.observationId,
          workingDirectory,
          exitCode: -1,
          stdout,
          stderr: error.message,
          toolName: options.toolName,
          reason: options.reason,
          startedAt,
          completedAt,
          success: false,
        });

        resolve({
          info,
          exitCode: -1,
          stdout,
          stderr: error.message,
        });
      });
    });
  }
}

/**
 * Run a command with the global client.
 */
export async function runCommand(
  command: string,
  options: RunCommandOptions & { client?: AgentTrace; traceId?: string } = {}
): Promise<RunCommandResult> {
  const { getClient, getCurrentTrace } = await import("./context");

  const client = options.client || getClient();
  if (!client) {
    throw new Error("No AgentTrace client available. Initialize one first.");
  }

  let traceId = options.traceId;
  if (!traceId) {
    const trace = getCurrentTrace();
    if (!trace) {
      throw new Error("No active trace. Create a trace first.");
    }
    traceId = trace.id;
  }

  const termClient = new TerminalClient(client);
  return termClient.run(traceId, command, options);
}

function generateId(): string {
  if (typeof crypto !== "undefined" && crypto.randomUUID) {
    return crypto.randomUUID();
  }
  return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c) => {
    const r = (Math.random() * 16) | 0;
    const v = c === "x" ? r : (r & 0x3) | 0x8;
    return v.toString(16);
  });
}
